File: MossTaps/golden-data/Original/P2 2014/Abramson, Christopher(11a3b6f00f76a75d9d650d008e4a510e)/Submission attachment(s)/Agent.py
# Your Agent for solving Raven's Progressive Matrices. You MUST modify this file.
#
# You may also create and submit new files in addition to modifying this file.
#
# Make sure your file retains methods with the signatures:
# def __init__(self)
# def Solve(self,problem)
#
# These methods will be necessary for the project's main method to run.

# Install Pillow and uncomment this line to access image processing.
from PIL import Image
from RavensFigure import RavensFigure
from RavensObject import RavensObject


# Define Attribute Scores for Weight Testing
ATTRIBUTE_SCORE = {'shape': 5,
                   'fill': 4,
                   'angle': 3,
                   'alignment': 3,
                   'size':2,
                   'inside': 1,
                   'above': 1}

# Define Alignment Scores For Weight Testing, First Object
ALIGN1 = {'bottom-right': (1, -1),
              'bottom-left': (-1, -1),
              'top-right': (1,1),
              'top-left': (-1, 1)
              }

# Define Alignment Scores For Weight Testing, Second Object
ALIGN2 = {(1, -1): 'bottom-right',
              (-1, -1): 'bottom-left',
              (1,1): 'top-right',
              (-1, 1): 'top-left'}

class Agent:
    # The default constructor for your Agent. Make sure to execute any
    # processing necessary before your Agent starts solving problems here.
    #
    # Do not add any variables to this signature; they will not be used by
    # main().
    def __init__(self):
        self.selections = []

    # The primary method for solving incoming Raven's Progressive Matrices.
    # For each problem, your Agent's Solve() method will be called. At the
    # conclusion of Solve(), your Agent should return an integer representing its
    # answer to the question: "1", "2", "3", "4", "5", or "6". These integers
    # are also the Names of the individual RavensFigures, obtained through
    # RavensFigure.getName() (as Strings).
    #
    # In addition to returning your answer at the end of the method, your Agent
    # may also call problem.checkAnswer(int givenAnswer). The parameter
    # passed to checkAnswer should be your Agent's current guess for the
    # problem; checkAnswer will return the correct answer to the problem. This
    # allows your Agent to check its answer. Note, however, that after your
    # agent has called checkAnswer, it will *not* be able to change its answer.
    # checkAnswer is used to allow your Agent to learn from its incorrect
    # answers; however, your Agent cannot change the answer to a question it
    # has already answered.
    #
    # If your Agent calls checkAnswer during execution of Solve, the answer it
    # returns will be ignored; otherwise, the answer returned at the end of
    # Solve will be taken as your Agent's answer to this problem.
    #
    # Make sure to return your answer *as an integer* at the end of Solve().
    # Returning your answer as a string may cause your program to crash.

    def is_match(self, case_start, case_end, sorted_origin, sorted_order, ans_match=False):

        record = []
        match = {}
        score_record = {}
        best_match, best_score = -1, -1
        for i in sorted_order:
            obj_dest = case_end.objects[i]
            for j in sorted_origin:
                obj_origin = case_start.objects[j]
                if (not ans_match and obj_origin not in record) or ans_match:
                    score = self.set_score(obj_origin, obj_dest)
                    if len(obj_origin.attributes) == len(obj_dest.attributes):
                        score += 1
                    if score > best_score:
                        best_score = score
                        best_match = obj_origin
                        match[obj_origin] = obj_dest
                        score_record[j] = best_score
                        record.append(obj_origin)

        remaining = set(case_start.objects.values()) - set(record)
        for obj in remaining:
            match[obj_origin] = None

        return match, score_record


    def set_score(self, object1, object2):
        score = 0
        for row, col in object2.attributes.iteritems():
            if row in object1.attributes and object1.attributes[row] == col:
                score += ATTRIBUTE_SCORE[row]
            elif row in ['inside', 'above', 'below'] and row in object1.attributes:
                score += 1
        return score


    def screen(self, rule, problem, case_start, case_end, case_x):
        if rule == 'number':
            counter_origin, counter_dest, counter_x = {}, {}, {}
            for figure, counter in zip([case_start, case_end, case_x], [counter_origin, counter_dest, counter_x]):
                for object in figure.objects:
                    shape = figure.objects[object].attributes['shape']
                    if shape not in counter:
                        counter[shape] = 1
                    else:
                        counter[shape] += 1
            predicted = {}
            sorted_order = sorted(counter_origin.keys(), key=lambda x: counter_origin[x], reverse=True)
            sorted_x = sorted(counter_x.keys(), key=lambda x: counter_x[x], reverse=True)
            helper = 0
            all_shapes = set(counter_origin.keys())|set(counter_dest.keys())
            delta_map = {}
            for key in all_shapes:
                if key not in counter_origin:
                    counter_origin[key] = 0
                if key not in counter_dest:
                    counter_dest[key] = 0
                delta_map[key] = counter_dest[key] - counter_origin[key]

            for shape1, shape2 in zip(sorted_order, sorted_x):
                dest_count = counter_dest.get(shape1) or 0
                delta = dest_count - counter_origin[shape1]
                predicted[shape2] = counter_x[shape2] + delta
                helper += 1

            if helper < len(sorted_x):
                for remaining in sorted_x[helper:]:
                    d = delta_map[remaining]
                    predicted[remaining] = counter_x[remaining] + d
            selections = []
            option_iterator = self.shape_count(problem)

            for i in xrange(1, 7):
                next_option = option_iterator.next()
                for shape in predicted:
                    no_match = False
                    if next_option.get(shape) != predicted[shape]:
                        no_match = True
                        break
                if not no_match:
                    selections.append(i)
            self.selections = selections
            return

        elif rule == 'detail' and len(case_start.objects) == len(case_x.objects):
            options = self.selections
            # get transform scores from objects have more attribute to less
            match = []
            sorted_origin = sorted(case_start.objects.keys(), key=lambda x:len(case_start.objects[x].attributes))
            for mp in [case_end, case_x]:
                sorted_order = sorted(mp.objects.keys(), key=lambda x:len(mp.objects[x].attributes))
                m, _ = self.is_match(case_start, mp, sorted_origin, sorted_order)
                match.append(m)

            # apply the transform rule to case_x
            other_match, case_x_match = match
            predicted = RavensFigure('predicted', '', '')
            helper = 0
            for start, case_x in case_x_match.iteritems():
                start_attrs = start.attributes
                if other_match[start]:
                    end_attrs = other_match[start].attributes
                else:
                    continue
                new_object = RavensObject(str(helper))

                for attr in end_attrs:
                    if attr == 'shape':
                        if start_attrs.get('shape') != end_attrs['shape']:
                            new_object.attributes['shape'] = end_attrs['shape']
                        else:
                            new_object.attributes['shape'] = case_x.attributes['shape']
                    elif attr == 'angle' and 'angle' in start_attrs and 'angle' in case_x.attributes:
                        # parallel relationship has higher order
                        if end_attrs['angle'] == start_attrs['angle']:
                            new_object.attributes['angle'] = case_x.attributes['angle']
                            continue
                        reflection = (int(end_attrs['angle']) + int(start_attrs['angle'])) % 360
                        if reflection == 0:
                            new_object.attributes['angle'] = str(360 - int(case_x.attributes['angle']))
                        elif reflection == 180:
                            angle = 180 - int(case_x.attributes['angle'])
                            if angle < 0:
                                angle += 360
                            new_object.attributes['angle'] = str(angle)
                        if start_attrs['angle'] == case_x.attributes['angle']:
                            new_object.attributes['angle'] = end_attrs['angle']
                    elif attr == 'alignment' and attr in start_attrs and attr in case_x.attributes:
                        alignment = [x['alignment'] for x in [start_attrs, end_attrs, case_x.attributes]]
                        p1, p2, p3 = [ALIGN1[x] for x in alignment]
                        if p1 == p3:
                            new_object.attributes['alignment'] = end_attrs['alignment']
                        elif p1[0] == p2[0] and p1[1] + p2[1] == 0:
                            p4 = (p3[0], -p3[1])
                            new_object.attributes['alignment'] = ALIGN2[p4]
                        elif p1[1] == p2[1] and p1[0] + p2[0] == 0:
                            p4 = (-p3[0], p3[1])
                            new_object.attributes['alignment'] = ALIGN2[p4]
                    else:
                        if start_attrs.get(attr) != end_attrs[attr]:
                            new_object.attributes[attr] = end_attrs[attr] if case_x.attributes.get(attr) == start_attrs.get(attr) else \
                            start_attrs.get(attr)
                        else:
                            new_object.attributes[attr] = case_x.attributes.get(attr)

                predicted.objects[str(helper)] = new_object
                helper += 1

            if self.selections:
                self.find_answer(predicted, problem, selections=True)
            else:
                self.find_answer(predicted, problem, selections=False)

        else:
            return

    def find_answer(self, predicted, problem, selections=False):
        if self.selections:
            options = [str(x) for x in self.selections]
        elif problem.problemType == '2x2':
            options = ['1', '2', '3', '4', '5', '6']
        max_score = 0
        candidate = None
        sorted_predict = sorted(predicted.objects.keys(), key=lambda x:len(predicted.objects[x].attributes))
        for i in options:
            figure = problem.figures[i]
            if len(figure.objects) != len(predicted.objects):
                continue
            sorted_order = sorted(figure.objects.keys(), key=lambda x:len(figure.objects[x].attributes))
            _, score = self.is_match(predicted, figure, sorted_predict, sorted_order, True)
            final_score = sum(score.values())
            if final_score > max_score:
                max_score = final_score
                candidate = int(i)
        self.selections = [candidate]
        return



    def shape_count(self, problem):
        if problem.problemType == '2x2':
            options = ['1', '2', '3', '4', '5', '6']
        for opt in options:
            counter = {}
            figure = problem.figures[opt]
            for i in figure.objects:
                object = figure.objects[i]
                if object.attributes['shape'] not in counter:
                    counter[object.attributes['shape']] = 1
                else:
                    counter[object.attributes['shape']] += 1
            yield counter


    def Solve(self,problem):
        self.selections = []
        select_order = ['number', 'detail']
        if problem.problemType == '2x2':
            case_start = problem.figures['A']
            case_end = problem.figures['B']
            case_x = problem.figures['C']
        if not problem.hasVerbal:
            return -1
        for rule in select_order:
            try:
                self.screen(rule=rule, problem=problem, case_start=case_start, case_end=case_end, case_x=case_x)
            except:
                continue
            if len(self.selections) == 1:
                return self.selections[0]
        if self.selections and len(self.selections) <= 2:
            return self.selections[-1]
        else:
            return -1
            
        
        
        
        

File: MossTaps/golden-data/Current/P2 2015/Newman, Joe(54fa6a0de05f3e459300f9e94f7228cf)/Submission attachment(s)/Agent.java
package ravensproject;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.TreeMap;
import java.util.stream.Collectors;

// Uncomment these lines to access image processing.
//import java.awt.Image;
//import java.io.File;
//import javax.imageio.ImageIO;

/**
 * Your Agent for solving Raven's Progressive Matrices. You MUST modify this
 * file.
 * 
 * You may also create and submit new files in addition to modifying this file.
 * 
 * Make sure your file retains methods with the signatures: public Agent()
 * public char Solve(RavensProblem problem)
 * 
 * These methods will be necessary for the project's main method to run.
 * 
 */
public class Agent {
	private final class ReverseComparator implements Comparator<Integer> {
		@Override
		public int compare(Integer o1, Integer o2) {
			return -o1.compareTo(o2);
		}
	}

	/**
	 * The default constructor for your Agent. Make sure to execute any
	 * processing necessary before your Agent starts solving problems here.
	 * 
	 * Do not add any variables to this signature; they will not be used by
	 * main().
	 * 
	 */
	public Agent() {

	}

	/**
	 * The primary method for solving incoming Raven's Progressive Matrices. For
	 * each problem, your Agent's Solve() method will be called. At the
	 * conclusion of Solve(), your Agent should return a String representing its
	 * answer to the question: "1", "2", "3", "4", "5", or "6". These Strings
	 * are also the Names of the individual RavensFigures, obtained through
	 * RavensFigure.getName().
	 * 
	 * In addition to returning your answer at the end of the method, your Agent
	 * may also call problem.checkAnswer(String givenAnswer). The parameter
	 * passed to checkAnswer should be your Agent's current guess for the
	 * problem; checkAnswer will return the correct answer to the problem. This
	 * allows your Agent to check its answer. Note, however, that after your
	 * agent has called checkAnswer, it will *not* be able to change its answer.
	 * checkAnswer is used to allow your Agent to learn from its incorrect
	 * answers; however, your Agent cannot change the answer to a question it
	 * has already answered.
	 * 
	 * If your Agent calls checkAnswer during execution of Solve, the answer it
	 * returns will be ignored; otherwise, the answer returned at the end of
	 * Solve will be taken as your Agent's answer to this problem.
	 * 
	 * @param problem
	 *            the RavensProblem your agent should solve
	 * @return your Agent's answer to this problem
	 */
	public int Solve(RavensProblem problem) {
//		if (!(problem.getName().equals("Basic Problem B-05") || problem.getName().equals("Basic Problem B-06"))) {
//		if (!problem.getName().equals("Basic Problem B-10")) {
//			return -1;
//		}
		// TODO fix wrong test problems
		
		if (!problem.hasVerbal()) {
			return -1;
		}
		System.out.print("Now solving " + problem.getName() + ".\t\t");
		Collection<ObjectWrapper> a;
		Collection<ObjectWrapper> b;
		Collection<ObjectWrapper> c;

		try {
			Collection<ObjectWrapper> expectedObjs = null;
			
			if (problem.getProblemType().equals("2x2")) {
				a = convertObjsInFigureToWrappers(problem.getFigures().get("A"));
				b = convertObjsInFigureToWrappers(problem.getFigures().get("B"));
				c = convertObjsInFigureToWrappers(problem.getFigures().get("C"));

				expectedObjs = solve2x2(a, b, c);
			} else {
				a = convertObjsInFigureToWrappers(problem.getFigures().get("G"));
				b = convertObjsInFigureToWrappers(problem.getFigures().get("H"));
				c = convertObjsInFigureToWrappers(problem.getFigures().get("H"));
				
				expectedObjs = solve3x3(a, b, c);
			}
			
			List<String> choiceIndexes;
			if (problem.getProblemType().equals("2x2")) {
				choiceIndexes = Arrays.asList("1", "2", "3", "4", "5", "6");
			} else {
				choiceIndexes = Arrays.asList("1", "2", "3", "4", "5", "6", "7", "8");
			}
			
			TreeMap<Integer, List<String>> choiceScores = new TreeMap<>(new ReverseComparator());
			HashMap<String, SemanticNetwork> choiceNetworks = new HashMap<>();
			
			for (String choiceIndex : choiceIndexes) {
				RavensFigure choiceFigure = problem.getFigures().get(choiceIndex);
				SemanticNetwork networkFromExpectedToChoice = buildNetwork(expectedObjs, convertObjsInFigureToWrappers(choiceFigure)); 
				choiceNetworks.put(choiceIndex, networkFromExpectedToChoice);
//				System.out.println("choice " + choiceIndex + ", network: " + networkFromExpectedToChoice);
				int choiceNetworkSimilarityScore = networkFromExpectedToChoice.getSimilarityScore();
				if (choiceScores.get(choiceNetworkSimilarityScore) == null) {
					choiceScores.put(choiceNetworkSimilarityScore, new ArrayList<>());
				}
				choiceScores.get(choiceNetworkSimilarityScore).add(choiceIndex);
			}

			String maxScoreChoiceIndex = "-1";

			// if more than 1 choice have the highest score, take positions into account
			List<String> highestScoreChoices = choiceScores.firstEntry().getValue();
			int currentLowestPositionDelta = 9999; 
			if (highestScoreChoices.size() > 1) {
				for (String choiceIndex : highestScoreChoices) {
					SemanticNetwork choiceNetwork = choiceNetworks.get(choiceIndex);
					int choiceNetworkDelta = choiceNetwork.getPositionDelta();
					if (choiceNetworkDelta < currentLowestPositionDelta) {
						currentLowestPositionDelta = choiceNetworkDelta; 
						maxScoreChoiceIndex = choiceIndex;
					}
				}
			} else {
				maxScoreChoiceIndex = highestScoreChoices.get(0); 
			}
			int givenAnswer = Integer.parseInt(maxScoreChoiceIndex);
			int correctAnswer = problem.checkAnswer(givenAnswer);
			String isCorrect = givenAnswer == correctAnswer ? "RIGHT" : "WRONG";
			System.out.println(isCorrect + ". Agent answer: " + givenAnswer + ". Correct answer: " + correctAnswer);
			return givenAnswer;
		} catch(Exception ex) {
			System.out.println("");
			System.out.println(ex);
			for (StackTraceElement stack : ex.getStackTrace()) {
				System.out.println(stack);
			}
		}
		System.out.println("Skipping " + problem.getName() + " because no answers found.");
		return -1;
	}

	private Collection<ObjectWrapper> solve3x3(Collection<ObjectWrapper> a, Collection<ObjectWrapper> b,
			Collection<ObjectWrapper> c) {
		SemanticNetwork abNetwork = buildNetwork(a, b);
		SemanticNetwork acNetwork = buildNetwork(a, c);
		
		return generateExpectedObjects(abNetwork, acNetwork, a, c);
	}

	private Collection<ObjectWrapper> solve2x2(Collection<ObjectWrapper> a, Collection<ObjectWrapper> b, Collection<ObjectWrapper> c) {
		// figure = each square/image
		// object = each part/shape in the image
		// attribute: of each part/shape
		
		SemanticNetwork abNetwork = buildNetwork(a, b);
		SemanticNetwork acNetwork = buildNetwork(a, c);
		
		Collection<ObjectWrapper> expectedObjs;
		
		// figure out whether to generate expected objects from B or C depending on which network contains more similar figures
		if (abNetwork.getSimilarityScore() >= acNetwork.getSimilarityScore()) {
			// A and B are more similar than A and C, so C should be the seed
			// for the missing figure
			expectedObjs = generateExpectedObjects(abNetwork, acNetwork, a, c);
		} else {
			expectedObjs = generateExpectedObjects(acNetwork, abNetwork, a, b);
		}
		return expectedObjs;
	}

	private SemanticNetwork buildNetwork(Collection<ObjectWrapper> srcObjs, Collection<ObjectWrapper> destObjs) {
		SemanticNetwork network = new SemanticNetwork();
		
		buildNetworkFromFigure(network, srcObjs);
		populatePositions(network, srcObjs);
		
		buildNetworkFromFigure(network, destObjs);
		populatePositions(network, destObjs);

		buildNetworkFromFigurePair(network, srcObjs, destObjs);
		
		return network;
	}

	@SuppressWarnings("unchecked")
	private Collection<ObjectWrapper> generateExpectedObjects(SemanticNetwork modelNetwork, SemanticNetwork correspondenceNetwork, 
			Collection<ObjectWrapper> a, Collection<ObjectWrapper> c) {
		// suppose A-B is the 'model' pair
		Collection<ObjectWrapper> expectedObjs = new ArrayList<>();
		for (ObjectWrapper objToTransform : c) {
			// find the corresponding object in A
			// the network should never have 2 edges from 2 different sources to the same destination, so `get(0)` is safe
			List<TransformationNetworkEdge> correspondingEdges = 
				correspondenceNetwork.getTransformationEdgesInList(correspondenceNetwork.findEdgesByDest(objToTransform));
			HashMap<String, String> expectedAttrs = (HashMap<String, String>) objToTransform.getAttributes().clone();
			boolean dontAdd = false;
			if (correspondingEdges.size() > 0) {
				ObjectWrapper correspondingObj = correspondingEdges.get(0).getSource();
				
				// find the transformation done to it
				List<TransformationNetworkEdge> edgesWithTransfToApply = modelNetwork.getTransformationEdgesOfObject(correspondingObj); 
				// apply the same transformation to objToTransform
				for (TransformationNetworkEdge edgeWithTransfToApply : edgesWithTransfToApply) {
					Transformation transformation = edgeWithTransfToApply.getTransformation();
					if (transformation instanceof ObjectDeleted) {
						dontAdd = true;
						break;
					}
					expectedAttrs = transformation.apply(expectedAttrs);
				}
			} 
			if (!dontAdd) {				
				// if no corresponding object found, just add as-is
				ObjectWrapper expectedObj = new ObjectWrapper(objToTransform);
				expectedObj.setAttributes(expectedAttrs);
				expectedObjs.add(expectedObj);
			}
		}
		
		// add any added object in the model network
		List<ObjectWrapper> addedObjs = modelNetwork.getTransformationEdges().stream().
				filter(edge -> edge.getTransformation() instanceof ObjectAdded).
				map(edge -> edge.getSource()).
				collect(Collectors.toList());
//		for (ObjectWrapper addedObj : addedObjs) {
//			// must be an added object whose relationship attributes are still pointing to their old parents
//			for (String relName : addedObj.getAttributes()) {
//				if (AttributeComparator.relationshipAttrs.contains(relName)) {
//					String origDests = addedObj.getAttributes().get(relName);
//					for (String dest : origDests.split(",")) {
//						// look in the model network first, if found, set attribute to corresponding objects
//						Optional<ObjectWrapper> origDest = modelNetwork.getTransformationDests().stream().filter(obj -> obj.getName().equals(dest)).findFirst();
//						if (origDest.isPresent()) {
//							modelNetwork.getTransDest(origDest.get());
//						}
//						// else, look in corr. network
//						
//					}
//					
//				}
//			}
//		}
		expectedObjs.addAll(addedObjs);
		
		// for each transformation from A to B, find the position delta between source and destination 
		for (ObjectWrapper srcObj : a) {
			List<TransformationNetworkEdge> transEdges = modelNetwork.getTransformationEdgesOfObject(srcObj);
			if (transEdges.size() > 0) {
				ObjectWrapper destObj = transEdges.get(0).getDestination();
				int rowDelta = destObj.getRow() - srcObj.getRow();
				int columnDelta = destObj.getColumn() - srcObj.getColumn();
				// apply the delta to the expected object generated from the object that corresponds to objToTransform 
				ObjectWrapper srcOfExpectedObj = correspondenceNetwork.getTransDest(srcObj);
				ObjectWrapper generatedObj = expectedObjs.stream().filter(object -> object.getName().equals(srcOfExpectedObj.getName())).findFirst().get();
				generatedObj.setRow(generatedObj.getRow() + rowDelta);
				generatedObj.setColumn(generatedObj.getColumn() + columnDelta);
			}
		}
		
		return expectedObjs;
	}

	private SemanticNetwork buildNetworkFromFigurePair(SemanticNetwork network, Collection<ObjectWrapper> srcObjs, Collection<ObjectWrapper> destObjs) {
		// this method should return a network with all edges filled in
		// to do that, while figuring out similarities, it needs to save all found similarities along with the total similarity score
		return buildNetworkFromObjLists(network, srcObjs, destObjs);
	}
	
	private SemanticNetwork buildNetworkFromObjLists(SemanticNetwork network, Collection<ObjectWrapper> srcObjList, Collection<ObjectWrapper> destObjList) {
		List<ObjectPairing> objPairings = Utilities.pairRelatedObjectsInFigures(network, srcObjList, destObjList);
		
		for (ObjectPairing pair : objPairings) {
			network.getEdges().addAll(pair.getNetworkEdges());
		}
		
		handleAddingAndDeleting(network, srcObjList, destObjList, objPairings);
		return network;
	}

	private Collection<ObjectWrapper> convertObjsInFigureToWrappers(RavensFigure source) {
		return source.getObjects().values().stream().map(object -> new ObjectWrapper(object)).collect(Collectors.toSet());
	}

	private void handleAddingAndDeleting(SemanticNetwork network, Collection<ObjectWrapper> srcObjList, Collection<ObjectWrapper> destObjList, List<ObjectPairing> objPairings) {
		// remove the sources (keys) of paired objects from the source figure. what remain are the deleted objects.  
		// remove the destinations (values) of paired objects from the destination figure. what remain are the added objects.  
		Collection<ObjectWrapper> deletedSrcObjs = new ArrayList<>(srcObjList);
		Collection<ObjectWrapper> addedDestObjs = new ArrayList<>(destObjList);
		Collection<ObjectWrapper> pairedSrcObjs = objPairings.stream().map(pair -> pair.getSrcObj()).collect(Collectors.toList());
		Collection<ObjectWrapper> pairedDestObjs = objPairings.stream().map(pair -> pair.getDestObj()).collect(Collectors.toList());
		// if there are objects in source that aren't in destination, create an OBJECT_ADDED edge
		deletedSrcObjs.removeAll(pairedSrcObjs);
		addedDestObjs.removeAll(pairedDestObjs);
		
		for (ObjectWrapper addedObj : addedDestObjs) {
			network.addEdge(new TransformationNetworkEdge(addedObj, addedObj, new ObjectAdded()));
		}
		// if there are objects in destination that aren't in source, create an OBJECT_DELETED edge 
		for (ObjectWrapper deletedObj : deletedSrcObjs) {
			network.addEdge(new TransformationNetworkEdge(deletedObj, deletedObj, new ObjectDeleted()));
		}
	}

	private void buildNetworkFromFigure(SemanticNetwork network, Collection<ObjectWrapper> objs) {
		// my end goal is to have a list of edges involving all objects in an island, in the order they're referred to, 
		// but not including non-adjacent edges. 
		// For example, if the edges are "a above b", "b above c", "a above c", I only want "a above b" and "b above c"
		// so, once I've figured out a is the starting object, since I have a list of objects ranked by connectedness, 
		// the next most connected object is b, so I search for an a-b edge and find it, so I add it to the 'linked list'
		// continue until the ranked list runs out

		for (ObjectWrapper obj : objs) {
			// if the object has a 'inside', 'above', etc attribute, create a
			// network edge from the object
			// to the other object in the figure with that key
			List<String> relAttrs = obj.getAttributes().keySet().stream().filter(attribute -> AttributeComparator.relationshipAttrs.contains(attribute)).collect(Collectors.toList());
			for (String relName : relAttrs) {
				// if a is inside b,c,d,e, split up, then add 1 edge per destination 
				String destinationsStr = obj.getAttributes().get(relName);
				String[] destinations = destinationsStr.split(",");
				for (String destination : destinations) {
					Optional<ObjectWrapper> destObjOptional = objs.stream().
						filter(object -> object.getName().equals(destination)).findFirst();
					ObjectWrapper destObject = destObjOptional.get();
					RelationshipNetworkEdge edge = new RelationshipNetworkEdge(obj, destObject, relName);
					network.addEdge(edge);
				}
			}
		}
	}

	private void populatePositions(SemanticNetwork network, Collection<ObjectWrapper> objs) {
		HashMap<String, TreeMap<Integer, ObjectWrapper>> numOfOutboundDests = new HashMap<>(); 

		for (ObjectWrapper obj : objs) {
			List<String> relAttrs = obj.getAttributes().keySet().stream().filter(attribute -> AttributeComparator.relationshipAttrs.contains(attribute)).collect(Collectors.toList());
			for (String relName : relAttrs) {
				if (numOfOutboundDests.get(relName) == null) {
					numOfOutboundDests.put(relName, new TreeMap<>(new ReverseComparator()));
				}
				// if a is inside b,c,d,e, split up, then add 1 edge per destination 
				String destinationsStr = obj.getAttributes().get(relName);
				String[] destinations = destinationsStr.split(",");
				numOfOutboundDests.get(relName).put(destinations.length, obj);
			}
		}
		// populate each object's row and column positions
		// for each relationship attribute found, find the starting object (e.g. for the 'above' attribute, the top object in 3 stacked objects)
		// which is the object with the most destinations
		// disclaimer: will fail if there are multiple 'islands' in the figure, e.g. a above b, c above d but no link between a/b and c/d. 
		// It'll never get to any island after the first one. 
		// compile a list of edges
		for (String relName : numOfOutboundDests.keySet()) {
			List<RelationshipNetworkEdge> linkedEdges = new LinkedList<>();
			TreeMap<Integer, ObjectWrapper> objsSortedByDestCount = numOfOutboundDests.get(relName);
			Entry<Integer, ObjectWrapper> startingObjEntry = objsSortedByDestCount.firstEntry();
			ObjectWrapper startingObj = startingObjEntry.getValue();
			// there are more than 1 object with this relationship, so we have a nested relationship on our hands
			// in which case we look at them 2 at a time to make the linked list of edges
			if (objsSortedByDestCount.size() > 1) {
				for (int i = 0; i < objsSortedByDestCount.size() - 1; i++) {
					Entry<Integer, ObjectWrapper> nextEntry = objsSortedByDestCount.higherEntry(startingObjEntry.getKey());
					ObjectWrapper nextObj = nextEntry.getValue();
					// find a relationship edge from the starting object to the next object in the list, having the current relationship
					List<RelationshipNetworkEdge> forwardEdges = network.getEdgesBySrcDestAndRel(startingObj, nextObj, relName);
					if (forwardEdges.size() > 0) {					
						linkedEdges.add(forwardEdges.get(0));
					} else {
						// chain is broken for some reason, probably shouldn't happen
						break;
					}
					startingObjEntry = nextEntry;
					startingObj = startingObjEntry.getValue();
				}
			}
			linkedEdges.add(
				network.getRelEdgesBySrcAndRel(startingObj, relName).get(0)
			);
			// then walk the edges 
			for (RelationshipNetworkEdge edge : linkedEdges) {
				ObjectWrapper srcObj = edge.getSource();
				ObjectWrapper destObj = edge.getDestination();
				
				switch (relName) {
				case RelationshipNetworkEdge.ABOVE:
					if (srcObj.getRow() == ObjectWrapper.INVALID_POSITION && destObj.getRow() == ObjectWrapper.INVALID_POSITION) {
						srcObj.setRow(0);
						destObj.setRow(1);
					} else if (srcObj.getRow() == ObjectWrapper.INVALID_POSITION) {
						srcObj.setRow(destObj.getRow() - 1);
					} else if (destObj.getRow() == ObjectWrapper.INVALID_POSITION)
						destObj.setRow(srcObj.getRow() + 1);
					break;
				case RelationshipNetworkEdge.LEFT_OF:
					if (srcObj.getColumn() == ObjectWrapper.INVALID_POSITION && destObj.getColumn() == ObjectWrapper.INVALID_POSITION) {
						srcObj.setColumn(0);
						destObj.setColumn(1);
					} else if (srcObj.getColumn() == ObjectWrapper.INVALID_POSITION) {
						srcObj.setColumn(destObj.getColumn() - 1);
					} else if (destObj.getColumn() == ObjectWrapper.INVALID_POSITION) {
						destObj.setColumn(srcObj.getColumn() + 1);
					}
					break;
				case RelationshipNetworkEdge.OVERLAPS:
					if (srcObj.getColumn() == ObjectWrapper.INVALID_POSITION && destObj.getColumn() == ObjectWrapper.INVALID_POSITION &&
						srcObj.getRow() == ObjectWrapper.INVALID_POSITION && destObj.getRow() == ObjectWrapper.INVALID_POSITION) {
						srcObj.setRow(0);
						srcObj.setColumn(0);
						destObj.setRow(0);
						destObj.setColumn(0);
					} 
					else {
						if (srcObj.getRow() == ObjectWrapper.INVALID_POSITION && destObj.getRow() != ObjectWrapper.INVALID_POSITION) {
							srcObj.setRow(destObj.getRow());
						} 
						if (srcObj.getRow() != ObjectWrapper.INVALID_POSITION && destObj.getRow() == ObjectWrapper.INVALID_POSITION) {
							destObj.setRow(srcObj.getRow());
						}
						if (srcObj.getColumn() == ObjectWrapper.INVALID_POSITION && destObj.getColumn() != ObjectWrapper.INVALID_POSITION) {
							srcObj.setColumn(destObj.getColumn());
						} 
						if (srcObj.getColumn() != ObjectWrapper.INVALID_POSITION && destObj.getColumn() == ObjectWrapper.INVALID_POSITION) {
							destObj.setColumn(srcObj.getColumn());
						}
					}
					break;
				default:
					break;
				}
			}
		}
		
		for (ObjectWrapper obj : objs) {
			if (obj.getRow() == ObjectWrapper.INVALID_POSITION) {
				obj.setRow(0);
			}
			if (obj.getColumn() == ObjectWrapper.INVALID_POSITION) {
				obj.setColumn(0);
			}
		}
	}
}
